### PR TITLE
feat: new module eventgrid /topic migrated from CARML to AVM

### PR DESCRIPTION
**Pull request by fabmas**
## Description

New module migrated from CARML to AVM

[avm.res.event-grid.topic](https://github.com/fabmas/bicep-registry-modules/actions/workflows/avm.res.event-grid.topic.yml)

### CODE REVIEW

File: avm/res/event-grid/topic/tests/e2e/waf-aligned/main.test.bicep [hunk 1]
```diff
@@ -0,0 +1,143 @@
+targetScope = 'subscription'
+
+metadata name = 'Using large parameter set'
+metadata description = 'This instance deploys the module with most of its features enabled.'
```
Comment (AlexanderSehr) [inline avm/res/event-grid/topic/tests/e2e/waf-aligned/main.test.bicep:4]:
```suggestion
metadata name = 'WAF-aligned'
metadata description = 'This instance deploys the module in alignment with the best-practices of the Azure Well-Architected Framework.'
```

Note: This requires an update of the readme

File: avm/res/event-grid/topic/tests/e2e/waf-aligned/main.test.bicep [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    managedIdentities: {
+      systemAssigned: true
+    }
+    roleAssignments: [
+      {
+        roleDefinitionIdOrName: 'Reader'
+        principalId: nestedDependencies.outputs.managedIdentityPrincipalId
+        principalType: 'ServicePrincipal'
+      }
+    ]
```
Comment (AlexanderSehr) [inline avm/res/event-grid/topic/tests/e2e/waf-aligned/main.test.bicep:134]:
```suggestion
```

Comment (AlexanderSehr) [reply avm/res/event-grid/topic/tests/e2e/waf-aligned/main.test.bicep:134]: Note: This requires an update of the readme

File: avm/res/event-grid/topic/tests/e2e/waf-aligned/main.test.bicep [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+  name: resourceGroupName
+  location: location
+}
+
+module nestedDependencies 'dependencies.bicep' = {
+  scope: resourceGroup
+  name: '${uniqueString(deployment().name, location)}-nestedDependencies'
+  params: {
+    virtualNetworkName: 'dep-${namePrefix}-vnet-${serviceShort}'
+    managedIdentityName: 'dep-${namePrefix}-msi-${serviceShort}'
```
Comment (AlexanderSehr) [inline avm/res/event-grid/topic/tests/e2e/waf-aligned/main.test.bicep:39]: May not be required with the below changes.

File: avm/res/event-grid/topic/tests/e2e/defaults/main.test.bicep [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        tags: {
+          'hidden-title': 'This is visible in the resource name'
+          Environment: 'Non-Prod'
+          Role: 'DeploymentValidation'
+        }
+      }
+    ]
+    managedIdentities: {
+      systemAssigned: true
+    }
```
Comment (AlexanderSehr) [inline avm/res/event-grid/topic/tests/e2e/defaults/main.test.bicep:73]: Are all of those parameters for the min test? I'd argue not.
If removed, plase also remove the dependencies & update the readme.

File: avm/res/event-grid/topic/main.bicep [hunk 5]
```diff
@@ -0,0 +1,350 @@
...
+
+type roleAssignmentType = {
+  @description('Required. The name of the role to assign. If it cannot be found you can specify the role definition ID instead.')
+  roleDefinitionIdOrName: string
+
+  @description('Required. The principal ID of the principal (user/group/identity) to assign the role to.')
+  principalId: string
+
+  @description('Optional. The principal type of the assigned principal ID.')
+  principalType: ('ServicePrincipal' | 'Group' | 'User' | 'ForeignGroup' | 'Device' | null)?
```
Comment (AlexanderSehr) [inline avm/res/event-grid/topic/main.bicep:231]:
```suggestion
  principalType: ('ServicePrincipal' | 'Group' | 'User' | 'ForeignGroup' | 'Device')?
```

File: avm/res/event-grid/topic/main.bicep [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+  }[]?
+
+  @description('Optional. The name of logs that will be streamed. "allLogs" includes all possible logs for the resource. Set to \'\' to disable log collection.')
+  metricCategories: {
+    @description('Required. Name of a Diagnostic Metric category for a resource type this setting is applied to. Set to \'AllMetrics\' to collect all metrics.')
+    category: string
+  }[]?
+
+  @description('Optional. A string indicating whether the export to Log Analytics should use the default destination type, i.e. AzureDiagnostics, or use a destination type.')
+  logAnalyticsDestinationType: ('Dedicated' | 'AzureDiagnostics' | null)?
```
Comment (AlexanderSehr) [inline avm/res/event-grid/topic/main.bicep:334]:
```suggestion
  logAnalyticsDestinationType: ('Dedicated' | 'AzureDiagnostics')?
```

**Review by skydiver1980 — APPROVED** [review]

File: avm/res/event-grid/topic/main.bicep [hunk 7]
```diff
@@ -0,0 +1,350 @@
...
+    ipConfigurations: privateEndpoint.?ipConfigurations
+    applicationSecurityGroupResourceIds: privateEndpoint.?applicationSecurityGroupResourceIds
+    customNetworkInterfaceName: privateEndpoint.?customNetworkInterfaceName
+  }
+}]
+
+resource topic_roleAssignments 'Microsoft.Authorization/roleAssignments@2022-04-01' = [for (roleAssignment, index) in (roleAssignments ?? []): {
+  name: guid(topic.id, roleAssignment.principalId, roleAssignment.roleDefinitionIdOrName)
+  properties: {
+    roleDefinitionId: contains(builtInRoleNames, roleAssignment.roleDefinitionIdOrName) ? builtInRoleNames[roleAssignment.roleDefinitionIdOrName] : roleAssignment.roleDefinitionIdOrName
```
Comment (AlexanderSehr) [inline avm/res/event-grid/topic/main.bicep:177]:
```suggestion
    roleDefinitionId: contains(builtInRoleNames, roleAssignment.roleDefinitionIdOrName) ? builtInRoleNames[roleAssignment.roleDefinitionIdOrName] : contains(roleAssignment.roleDefinitionIdOrName, '/providers/Microsoft.Authorization/roleDefinitions/') ? roleAssignment.roleDefinitionIdOrName : subscriptionResourceId('Microsoft.Authorization/roleDefinitions', roleAssignment.roleDefinitionIdOrName)
```

We're currently introducing this change across the board

File: avm/res/event-grid/topic/tests/e2e/max/main.test.bicep [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+          Role: 'DeploymentValidation'
+        }
+      }
+    ]
+    roleAssignments: [
+      {
+        roleDefinitionIdOrName: 'Reader'
+        principalId: nestedDependencies.outputs.managedIdentityPrincipalId
+        principalType: 'ServicePrincipal'
+      }
```
Comment (AlexanderSehr) [inline avm/res/event-grid/topic/tests/e2e/max/main.test.bicep:135]:
```suggestion
      {
        roleDefinitionIdOrName: 'Owner'
        principalId: nestedDependencies.outputs.managedIdentityPrincipalId
        principalType: 'ServicePrincipal'
      }
      {
        roleDefinitionIdOrName: 'b24988ac-6180-42a0-ab88-20f7382dd24c'
        principalId: nestedDependencies.outputs.managedIdentityPrincipalId
        principalType: 'ServicePrincipal'
      }
      {
        roleDefinitionIdOrName: subscriptionResourceId('Microsoft.Authorization/roleDefinitions', 'acdd72a7-3385-48ef-bd42-f606fba81ae7')
        principalId: nestedDependencies.outputs.managedIdentityPrincipalId
        principalType: 'ServicePrincipal'
      }
```

File: avm/res/event-grid/topic/tests/e2e/max/main.test.bicep [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+      kind: 'CanNotDelete'
+      name: 'myCustomLockName'
+    }
+    privateEndpoints: [
+      {
+        privateDnsZoneResourceIds: [
+          nestedDependencies.outputs.privateDNSZoneResourceId
+        ]
+        service: 'topic'
+        subnetResourceId: nestedDependencies.outputs.subnetResourceId
```
Comment (AlexanderSehr) [inline avm/res/event-grid/topic/tests/e2e/max/main.test.bicep:122]:
```suggestion
        subnetResourceId: nestedDependencies.outputs.subnetResourceId
        roleAssignments: [
          {
            roleDefinitionIdOrName: 'Owner'
            principalId: nestedDependencies.outputs.managedIdentityPrincipalId
            principalType: 'ServicePrincipal'
          }
          {
            roleDefinitionIdOrName: subscriptionResourceId('Microsoft.Authorization/roleDefinitions', 'acdd72a7-3385-48ef-bd42-f606fba81ae7')
            principalId: nestedDependencies.outputs.managedIdentityPrincipalId
            principalType: 'ServicePrincipal'
          }
        ]
```

File: avm/res/event-grid/topic/tests/e2e/max/main.test.bicep (hunk 8 above)
Comment (AlexanderSehr) [reply avm/res/event-grid/topic/tests/e2e/max/main.test.bicep:135]: Please remember to update the readme after applying all changes

**Review by AlexanderSehr — APPROVED** [review]

**Pull request merged**